orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RollingRedBlackStrategy.groovy

```diff
@@ -86,72 +90,118 @@ class RollingRedBlackStrategy implements Strategy, ApplicationContextAware {
-          source: getSource(targetServerGroupResolver, stageData, baseContext),
-          targetLocation: cleanupConfig.location,
-          scalePct: p,
-          pinCapacity: p < 100 // if p = 100, capacity should be unpinned
+        target        : TargetServerGroup.Params.Target.current_asg_dynamic,
+        action        : ResizeStrategy.ResizeAction.scale_to_server_group,
+        source        : source,
+        targetLocation: cleanupConfig.location,
+        scalePct      : p,
+        pinCapacity   : p < 100 // if p = 100, capacity should be unpinned
       ]
 
-      stages << newStage(
-          stage.execution,
-          resizeServerGroupStage.type,
-          "Grow to $p% Desired Size",
-          resizeContext,
-          stage,
-          SyntheticStageOwner.STAGE_AFTER
+      def resizeStage = newStage(
+        stage.execution,
+        resizeServerGroupStage.type,
+        "Grow to $p% Desired Size",
+        resizeContext,
+        stage,
+        SyntheticStageOwner.STAGE_AFTER
       )
+      stages << resizeStage
+
+      // an expression to grab newly deployed server group at runtime (ie. the server group being resized up)
+      def deployedServerGroupName = '${' + "#stage('${resizeStage.id}')['context']['asgName']" + '}'.toString()
+      stages.addAll(getBeforeCleanupStages(stage, stageData, source, deployedServerGroupName, p))
 
       def disableContext = baseContext + [
-          desiredPercentage           : p,
-          remainingEnabledServerGroups: 1,
-          preferLargerOverNewer       : false
+        desiredPercentage           : p,
+        remainingEnabledServerGroups: 1,
+        preferLargerOverNewer       : false
       ]
 
-      if(stageData?.delayBeforeDisableSec) {
-        def waitContext = [waitTime: stageData?.delayBeforeDisableSec]
-        stages << newStage(
-          stage.execution,
-          waitStage.type,
-          "wait",
-          waitContext,
-          stage,
-          SyntheticStageOwner.STAGE_AFTER
-        )
-      }
-
-      // TODO-AJ this has questionable behavior w/ > 1 old server group
       stages << newStage(
-          stage.execution,
-          disableClusterStage.type,
-          "Disable $p% of Traffic",
-          disableContext,
-          stage,
-          SyntheticStageOwner.STAGE_AFTER
+        stage.execution,
+        disableClusterStage.type,
+        "Disable $p% of Traffic",
+        disableContext,
+        stage,
+        SyntheticStageOwner.STAGE_AFTER
       )
-
     })
 
     return stages
   }
 
+  List<Stage> getBeforeCleanupStages(Stage parentStage,
+                                     RollingRedBlackStageData stageData,
+                                     ResizeStrategy.Source source,
+                                     String deployedServerGroupName,
+                                     int percentageComplete) {
+    def stages = []
+
+    if (stageData.getDelayBeforeCleanup()) {
```

Comment:
Not likely you would have both a delay + pipeline, I plan on preventing it in the UI.